test/xbgpu/test_bsend.py

```diff
@@ -0,0 +1,278 @@
+        # send_heap just queues data for sending but is non-blocking.
+        # Flush to ensure that the data all gets sent before we return.
+        await send_stream.stream.async_flush()
+
+        return data
+
+    @staticmethod
+    async def _recv_data(
+        data: np.ndarray,
+        queues: list[spead2.InprocQueue],
+        n_engines: int,
+        engine_id: int,
+        channel_offset: int,
+        n_channels_per_substream: int,
+        n_spectra_per_heap: int,
+        heap_timestamp_step: int,
+    ) -> None:
+        """Receive data transmitted from :meth:`_send_data`.
+
+        Error-check data here as well.
+
+        Parameters
+        ----------
+        data
+            Random data generated during data transmission, of shape
+            - (N_TX_ITEMS, HEAPS_PER_CHUNK, len(outputs), n_channels_per_substream, n_spectra_per_heap, COMPLEX)
+        queues
+            List of :class:`spead2.InprocQueue` used to transmit heaps
+            in :meth:`_send_data`.
+        n_engines, engine_id, channel_offset, n_channels_per_substream, n_spectra_per_heap, heap_timestamp_step
+            Variables declared by the calling unit test to verify
+            transmitted data.
+        """
+        # Reshape as we verify *heaps* per substream, not chunks
+        data = data.reshape((TX_HEAPS_PER_SUBSTREAM,) + data.shape[2:])
+
+        out_config = spead2.recv.StreamConfig()
+        out_tp = spead2.ThreadPool()
+        for i, queue in enumerate(queues):
+            stream = spead2.recv.asyncio.Stream(out_tp, out_config)
+            stream.add_inproc_reader(queue)
+
+            # Wait for the first packet to arrive - it is expected to be the
+            # SPEAD descriptor.
+            ig = spead2.ItemGroup()
+            heap = await stream.get()
+            assert heap.cnt % n_engines == engine_id, "The heap IDs are not correctly strided"
+            items = ig.update(heap)
+            assert items == {}, "This heap contains item values not just the expected descriptors."
+
+            # Check the data heaps
+            for j in range(TX_HEAPS_PER_SUBSTREAM):
+                heap = await stream.get()
+                items = ig.update(heap)
+                assert set(items.keys()) == {"timestamp", "frequency", "bf_raw"}
+                assert items["timestamp"].id == TIMESTAMP_ID
+                assert items["timestamp"].value == j * heap_timestamp_step
+                assert items["frequency"].id == FREQUENCY_ID
+                assert items["frequency"].value == channel_offset
+                assert items["bf_raw"].id == BF_RAW_ID
+                assert items["bf_raw"].value.shape == (n_channels_per_substream, n_spectra_per_heap, COMPLEX)
+                assert items["bf_raw"].value.dtype == np.int8
+                np.testing.assert_equal(items["bf_raw"].value, data[j, i, ...])
+
+    @pytest.mark.combinations(
+        "num_ants, num_channels, num_spectra_per_heap",
+        [1, 19, 80],
```

Comment:
It still seems very roundabout to parametrise the test by num_ants and then use a completely arbitrary formula to determine an n_engines value from that. Why not just parametrise n_engines directly e.g.
```suggestion
        "n_engines, num_channels, num_spectra_per_heap",
        [4, 128, 512],
```
Apart from anything else, it would make the test identifier actually contain the number of engines being tested.

Reply:
Ah, I see what you were referring to in the previous comment. Shall do, shall do.